algorithms/linfa-clustering/src/dbscan/hyperparameters.rs

```diff
@@ -7,54 +8,70 @@ use serde_crate::{Deserialize, Serialize};
     derive(Serialize, Deserialize),
     serde(crate = "serde_crate")
 )]
-#[derive(Clone, Debug, PartialEq)]
+#[derive(Debug)]
 /// The set of hyperparameters that can be specified for the execution of
 /// the [DBSCAN algorithm](struct.Dbscan.html).
-pub struct DbscanHyperParams<F: Float> {
+pub struct DbscanHyperParams<F: Float, D: Distance<F>, N: NearestNeighbour> {
     /// Distance between points for them to be considered neighbours.
     tolerance: F,
     /// Minimum number of neighboring points a point needs to have to be a core
     /// point and not a noise point.
     min_points: usize,
+    /// Distance metric used in the DBSCAN calculation
+    dist_fn: D,
+    /// Nearest neighbour algorithm used for range queries
+    nn_impl: N,
 }
 
 /// Helper struct used to construct a set of hyperparameters for [DBSCAN algorithm](struct.Dbscan.html).
-pub struct DbscanHyperParamsBuilder<F: Float> {
-    tolerance: F,
-    min_points: usize,
+pub struct DbscanHyperParamsBuilder<F: Float, D: Distance<F>, N: NearestNeighbour> {
+    pub(crate) tolerance: F,
+    pub(crate) min_points: usize,
+    pub(crate) dist_fn: D,
+    pub(crate) nn_impl: N,
 }
 
-impl<F: Float> DbscanHyperParamsBuilder<F> {
+impl<F: Float, D: Distance<F>, N: NearestNeighbour> DbscanHyperParamsBuilder<F, D, N> {
     /// Distance between points for them to be considered neighbours.
     pub fn tolerance(mut self, tolerance: F) -> Self {
         self.tolerance = tolerance;
         self
     }
 
+    /// Nearest neighbour algorithm used for range queries
+    pub fn nn_impl(mut self, nn_impl: N) -> Self {
+        self.nn_impl = nn_impl;
+        self
+    }
+
+    /// Distance metric used in the DBSCAN calculation
+    pub fn dist_fn(mut self, dist_fn: D) -> Self {
```

Comment:
for `fn` it's a bit more common, but why not `dist_metric`?

Reply:
`dist_fn` is already used everywhere in `linfa-nn` so I want to keep it consistent. If `dist_fn` is confusing I can change it to `dist_metric` across both crates.

Reply:
yeah `dist_fn` is obvious enough